addons/project/controllers/portal.py

```diff
@@ -356,17 +360,54 @@ def _prepare_tasks_values(self, page, date_begin, date_end, sortby, search, sear
         if search and search_in:
             domain += self._task_get_search_domain(search_in, search)
 
+        milestone_domain = AND([domain, [('allow_milestones', '=', 'True')]])
+        milestones_allowed = TaskSudo.search_count(milestone_domain, limit=1) == 1
+
+        searchbar_sortings = dict(sorted(self._task_get_searchbar_sortings(milestones_allowed).items(),
+                                         key=lambda item: item[1]["sequence"]))
+        searchbar_inputs = self._task_get_searchbar_inputs(milestones_allowed)
+        searchbar_groupby = self._task_get_searchbar_groupby(milestones_allowed)
+
+        # default sort by value
+        if not sortby or (sortby == 'milestone' and not milestones_allowed):
+            sortby = 'date'
+        order = searchbar_sortings[sortby]['order']
+
+        # default group by value
+        if not groupby or (groupby == 'milestone' and not milestones_allowed):
+            groupby = 'project'
+
+        # search reset if needed
+        if not milestones_allowed:
+            if search_in == 'milestone':
+                search_in = 'all'
+
         # content according to pager and archive selected
         order = self._task_get_order(order, groupby)
 
         def get_grouped_tasks(pager_offset):
-            tasks = Task.search(domain, order=order, limit=self._items_per_page, offset=pager_offset)
+            tasks = TaskSudo.search(domain, order=order, limit=self._items_per_page, offset=pager_offset)
             request.session['my_project_tasks_history' if url.startswith('/my/projects') else 'my_tasks_history'] = tasks.ids[:100]
 
+            tasks_project_allow_milestone = tasks.filtered(lambda t: t.allow_milestones)
+            tasks_no_milestone = tasks - tasks_project_allow_milestone
+
             groupby_mapping = self._task_get_groupby_mapping()
             group = groupby_mapping.get(groupby)
             if group:
-                grouped_tasks = [Task.concat(*g) for k, g in groupbyelem(tasks, itemgetter(group))]
+                if group == 'milestone_id':
+                    grouped_tasks = [Task.concat(*g) for k, g in groupbyelem(tasks_project_allow_milestone, itemgetter(group))]
+
+                    if not grouped_tasks:
+                        grouped_tasks = [tasks_no_milestone]
+                    else:
+                        if grouped_tasks[len(grouped_tasks) - 1][0].milestone_id and tasks_no_milestone:
+                            grouped_tasks.append(tasks_no_milestone)
+                        else:
+                            grouped_tasks[len(grouped_tasks) - 1] |= tasks_no_milestone
```

Comment:
As far as i've tested, yes. But maybe the testing was not made with enough data, and in some cases, it wont. Is it better to make some sort of search/check to be sure of where the 'no_milestone' task in the grouped task ?

Reply:
Perhaps the functions used already sort the data some it should be good but since I don't know those functions, I am not sure. If you are sure then it is okay for me.